platform/mbed_error.c

```diff
@@ -188,24 +265,97 @@ WEAK MBED_NORETURN mbed_error_status_t mbed_error(mbed_error_status_t error_stat
         (void) handle_error(error_status, error_value, filename, line_number, MBED_CALLER_ADDR());
 
         //On fatal errors print the error context/report
-        ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number);
+        ERROR_REPORT(&last_error_ctx, error_msg, filename, line_number, true);
+    }
+    
+#if MBED_CONF_PLATFORM_CRASH_CAPTURE_ENABLED
+    uint32_t crc_val = 0;
+    crc_val = compute_crc32( (unsigned char *)report_error_ctx, ((uint32_t)&(report_error_ctx->crc_error_ctx) - (uint32_t)report_error_ctx) );
+    //Read report_error_ctx and check if CRC is correct for report_error_ctx
+    if(report_error_ctx->crc_error_ctx == crc_val) {
+        uint32_t current_reboot_count = report_error_ctx->error_reboot_count;
+        last_error_ctx.error_reboot_count = current_reboot_count + 1;
+    } else {
+        last_error_ctx.error_reboot_count = 1;
     }
+    last_error_ctx.is_error_processed = 0;//Set the flag that this is a new error
+    //Update the struct with crc
+    last_error_ctx.crc_error_ctx = compute_crc32( (unsigned char *)&last_error_ctx, ((uint32_t)&(last_error_ctx.crc_error_ctx) - (uint32_t)&last_error_ctx) );
+    //Protect report_error_ctx while we update it
+    core_util_critical_section_enter();
+    memcpy(report_error_ctx, &last_error_ctx, sizeof(mbed_error_ctx));
+    core_util_critical_section_exit();
+    //We need not call delete_mbed_crc(crc_obj) here as we are going to reset the system anyway, and calling delete while handling a fatal error may cause nested exception
+#if MBED_CONF_PLATFORM_FATAL_ERROR_AUTO_REBOOT_ENABLED
+    system_reset();//do a system reset to get the system rebooted
+    while(1);
```

Comment:
while(1) is dead here. system_reset is MBED_NORETURN.

I would be strongly inclined to have an exponential backoff delay on each reboot. 5 seconds first time, say, and double each time.

Reply:
while(1) was an attempt to stop the system in case system_reset() ever changed behavior. But I think its confusing, so will remove it.

I initially thought of having delay before reboot, but later I decided it wont help much, unless we want to provide opportunity to attach a debugger or something, is that why we should add the backoff delay?

Reply:
It's more a stability principle - each time the device boots, it may well be using extra network resources to reinitialise itself.  The problem leading to crash may even be exacerbated by overall system load caused by other rebooting devices. A network could collapse as a result.

A general networking principle is that retries should be exponentially backed off to guarantee that the network is stable, and doesn't undergo congestion collapse. (Many protocols and standards use parameters based on chapter 14 of [RFC 3315](https://tools.ietf.org/html/rfc3315) - initial time, maximum time, maximum count...)

Obviously we would require someone at some point to reset the count, and hence the backoff, but you've left that application dependent. Were that to be reset too readily, there might be a problem.

In this case, maybe you don't want to literally double the "delay after crashing before rebooting", but rather double the "minimum time since we last rebooted". Not sure what your best timer for that would be - can't safely use `Kernel::get_ms_count`, so might need to have a boot-started `LowPowerTimer'? 

If you're initially configuring the count very low though, this is less necessary - you'd want this with high or unlimited counts. So it's maybe an extension.